### PR TITLE
Mariacarlina/new version

### PR DESCRIPTION
**Pull request by mariacarlinahernandez**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
+
 
-    
-## Send one value to Ubidots
+## Change the default name
 
-To send a value to Ubidots, go to **Sketch -> Examples -> UbidotsArduino library** and select the "saveValue" example. 
-Put your Ubidots token and variable ID where indicated, as well as the WiFi settings.
-Upload the code, open the Serial monitor to check the results. If no response is seen, try unplugging your Arduino and then plugging it again. Make sure the baud rate of the Serial monitor is set to the same one specified in your code.
+The library will create a new Ubidots device named **"Arduino-Ethernet"**, also assigns the same name for the device label. If you desire assign a different device label, you can add the line below into the setup:
```
Comment (jotathebest) [inline README.md]: Correct line 32, 'if you desire to assign a differente device label, please add to your `setup()` function the line below

File: README.md [hunk 2]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
+char const * VARIABLE_LABEL_2 = "humidity"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_3 = "pressure"; // Assign the unique variable label to send the data
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
 byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
-// Set the static IP address to use if the DHCP fails to assign
+
+/* Set the static IP address to use if the DHCP fails to assign */
 IPAddress ip(192, 168, 0, 177);
```
Comment (jotathebest) [inline README.md]: Set here the google's DNS, 8.8.8.8

File: README.md [hunk 3]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
+ *******************************/
+/* Assigns the Ubidots parameters */
+char const * TOKEN = "Assign_your_Ubidots_TOKEN_here"; // Assign your Ubidots TOKEN
+char const * VARIABLE_LABEL_1 = "temperature"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_2 = "humidity"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_3 = "pressure"; // Assign the unique variable label to send the data
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
 byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
```
Comment (jotathebest) [inline README.md:65]: @mariacarlinahernandez please add a method that extracts the MAC inside our library

File: README.md [hunk 4]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
+char const * VARIABLE_LABEL_2 = "humidity"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_3 = "pressure"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_4 = "speed"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_5 = "current"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_6 = "voltage"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_7 = "light"; // Assign the unique variable label to send the data
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
 byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
```
Comment (jotathebest) [inline README.md:129]: Add a method to retrieve the MAC

File: README.md [hunk 5]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
-Upload the code, open the Serial monitor to check the results. If no response is seen, try unplugging your Arduino and then plugging it again. Make sure the baud rate of the Serial monitor is set to the same one specified in your code.
+## Get the last value from Ubidots
+
+To get the last value of a variable from Ubidots, go to **Sketch -> Examples -> UbidotsArduino library** and select the **"UbidotsGetValue"** example.
+
+Assign your Ubidots [TOKEN](http://help.ubidots.com/user-guides/find-your-token-from-your-ubidots-account), the device and variable label from where you desire obtain the value, also the MAC address for your controller.
+
+Upload the code, open the Serial monitor to check the results.
+
+**NOTE**: ging it again. Make sure the baud rate of the Serial monitor is set to the same one specified in your code.
```
Comment (jotathebest) [inline README.md]: What is 'ging' ?

File: README.md [hunk 6]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
+char const * DEVICE_LABEL = "Assign_device_label_here"; // Assign the unique device label
+char const * VARIABLE_LABEL = "Assign_variable_label_here"; // Assign the unique variable label to get the last value
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
 byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
-// Set the static IP address to use if the DHCP fails to assign
+
+/* Set the static IP address to use if the DHCP fails to assign */
 IPAddress ip(192, 168, 0, 177);
```
Comment (jotathebest) [inline README.md]: put google's DNS

File: examples/UbidotsGetValue/UbidotsGetValue.ino [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ * Constants and objects
+ *******************************/
+/* Assigns the Ubidots parameters */
+char const * TOKEN = "Assign_your_Ubidots_TOKEN_here"; // Assign your Ubidots TOKEN
+char const * DEVICE_LABEL = "Assign_device_label_here"; // Assign the unique device label
+char const * VARIABLE_LABEL = "Assign_variable_label_here"; // Assign the unique variable label to get the last value
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
+byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
```
Comment (jotathebest) [inline examples/UbidotsGetValue/UbidotsGetValue.ino:18]: IDEM to previous Review

File: examples/UbidotsGetValue/UbidotsGetValue.ino [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+/* Set the static IP address to use if the DHCP fails to assign */
+IPAddress ip(192, 168, 0, 177);
```
Comment (jotathebest) [inline examples/UbidotsGetValue/UbidotsGetValue.ino]: IDEM to previous Review

File: examples/UbidotsGetValue/UbidotsGetValue.ino [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+Ubidots client(TOKEN);
+
+/********************************
+ * Main Functions
+ *******************************/
+void setup() {
+    Serial.begin(9600);
+    //client.setDebug(true);// uncomment this line to visualize the debug message
+    /* start the Ethernet connection */
+    if (Ethernet.begin(mac) == 0) {
```
Comment (jotathebest) [inline examples/UbidotsGetValue/UbidotsGetValue.ino]: This should be a loop, what happens if the ethernet cannot be initialized?

File: examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+char const * VARIABLE_LABEL_2 = "humidity"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_3 = "pressure"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_4 = "speed"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_5 = "current"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_6 = "voltage"; // Assign the unique variable label to send the data
+char const * VARIABLE_LABEL_7 = "light"; // Assign the unique variable label to send the data
+
+/* Enter a MAC address for your controller below */
+/* Newer Ethernet shields have a MAC address printed on a sticker on the shield */
+byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };
```
Comment (jotathebest) [inline examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino:23]: IDEM to previous Review

File: examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+/* Set the static IP address to use if the DHCP fails to assign */
+IPAddress ip(192, 168, 0, 177);
```
Comment (jotathebest) [inline examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino]: IDEM to previous Review

File: examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+/********************************
+ * Main Functions
+ *******************************/
+void setup() {
+    Serial.begin(9600);
+    client.setDebug(true);
+    //client.setDeviceLabel("my-new-device");// uncomment this line to change the defaul label
+
+    /* start the Ethernet connection */
+    if (Ethernet.begin(mac) == 0) {
```
Comment (jotathebest) [inline examples/UbidotsSaveMultiValues/UbidotsSaveMultiValues.ino]: IDEM to previous Review

File: src/UbidotsEthernet.cpp [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+    the function returns ERRORs messages
+ */
+float Ubidots::getValue(char* device_label, char* variable_label) {
+  /* Assigns the constans as global on the function */
+  char* res = (char *) malloc(sizeof(char) * 250);
+  char* response;
+  char* serverResponse;
+  float num;
+  String resp_str = "";
+  uint8_t i = 0; // PREGUNTAR SI SE PUEDEN ALMACENAR EN UNO SOLO
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: Erase the spanish comment

File: src/UbidotsEthernet.cpp [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+
+  /* Reads the response from the server */
+  while (_client.available()) {
+    char c = _client.read();
+    if (c == -1) {
+      if (_debug) {
+        Serial.println(F("Error reading data from server"));
+      }
+      _client.stop();
+      delay(5);
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: is this delay necessary?

File: src/UbidotsEthernet.cpp [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+ * @return num the data that you get from the Ubidots API, if any error occurs
+    the function returns ERRORs messages
+ */
+float Ubidots::getValue(char* device_label, char* variable_label) {
+  /* Assigns the constans as global on the function */
+  char* res = (char *) malloc(sizeof(char) * 250);
+  char* response;
+  char* serverResponse;
+  float num;
+  String resp_str = "";
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: Do not use Strings

File: src/UbidotsEthernet.cpp [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+
+  /* Converts the response to a char */
+  resp_str.toCharArray(res, 250);
+  /* Parses the response to get just the last value received */
+  response = strtok(res, "\r\n");
+  while(response!=NULL) {
+    j++;
+    printf("%s", response);
+    response = strtok(NULL, "\r\n");
+    if (j == 10) {
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: why 10? Please make a comment saying the reason to check when it is 10

File: src/UbidotsEthernet.cpp [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+    _deviceLabel = new_device_label;
+}
+
+/**
+ * Gets the length of the body
+ * @arg body data to be send - JSON
+ * @return dataLen the length of the JSON
+ */
+int Ubidots::dataLen(char* body) {
+  int dataLen = 0;
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: this should be uint8_t

File: src/UbidotsEthernet.cpp [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+
+/**
+ * Send all data of all variables that you saved
+ * @reutrn true upon success, false upon error.
+ */
+bool Ubidots::sendAll() {
+  /* Assigns the constans as global on the function */
+  uint8_t max_retries = 0;
+  int timeout = 0;
+  int i = 0;
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: this should be uint8_t

File: src/UbidotsEthernet.cpp [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+      delay(5);
+      free(data);
+      return ERROR_VALUE;
+    }
+  }
+
+  currentValue = 0;
+  free(data);
+  free(body);
+  //memset(data, 0, sizeof(data));
```
Comment (jotathebest) [inline src/UbidotsEthernet.cpp]: remove these commented lines

File: src/UbidotsEthernet.h [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    void add(char *variable_label, double value, char *ctext, unsigned long timestamp);
+    void setDebug(bool debug);
+    void setDeviceLabel(char* new_device_label);
+
+  private:
+    bool _debug = false;
+    char* _deviceLabel;
+    char* _token;
+    char* _server;
+    int dataLen(char* body);
```
Comment (jotathebest) [inline src/UbidotsEthernet.h]: This should be a variable to store the length of the char array, right now it is not doing anything in this declaration

File: README.md [hunk 21]
```diff
@@ -11,142 +9,226 @@ We can't guarantee that it will work with different Ethernet modules.
...
     Serial.begin(9600);
-    // start the Ethernet connection:
+    client.setDebug(true);
+    //client.setDeviceLabel("my-new-device");// uncomment this line to change the defaul label
+
+    /* start the Ethernet connection */
     if (Ethernet.begin(mac) == 0) {
       Serial.println("Failed to configure Ethernet using DHCP");
-      // try to congifure using IP address instead of DHCP:
+      /* Try to congifure using IP address instead of DHCP */
```
Comment (jotathebest) [inline README.md]: 'configure'

File: src/UbidotsEthernet.cpp [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+ */
+bool Ubidots::sendAll() {
+  /* Assigns the constans as global on the function */
+  uint8_t max_retries = 0;
+  uint8_t timeout = 0;
+  uint8_t i = 0;
+  String str;
+
+  /* Saves variable value in str */
+  str = String(((val+i)->varValue),3); // variable value
```
Comment (woakas) [inline src/UbidotsEthernet.cpp]: @mariacarlinahernandez posible overflow, what happen when `val` has no values and the user invoke sendAll?

File: src/UbidotsEthernet.cpp [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+ * @reutrn true upon success, false upon error.
+ */
+bool Ubidots::sendAll() {
+  /* Assigns the constans as global on the function */
+  uint8_t max_retries = 0;
+  uint8_t timeout = 0;
+  uint8_t i = 0;
+  String str;
+
+  /* Saves variable value in str */
```
Comment (woakas) [inline src/UbidotsEthernet.cpp]: if (currentValue == 0) {
    return null;
}

**Pull request merged**